### PR TITLE
EksaRelease CRD

### PR DESCRIPTION
**Pull request by tatlat**
*Issue #, if available:*

#5847 

*Description of changes:*
Added EKSARelease type and CRD definition

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -422,7 +422,7 @@ unit-test: ## Run unit tests
 unit-test: $(SETUP_ENVTEST) 
 unit-test: KUBEBUILDER_ASSETS ?= $(shell $(SETUP_ENVTEST) use --use-env -p path --arch $(GO_ARCH) $(KUBEBUILDER_ENVTEST_KUBERNETES_VERSION))
 unit-test:
-	KUBEBUILDER_ASSETS="$(KUBEBUILDER_ASSETS)" $(GO_TEST) $(UNIT_TEST_PACKAGES) -cover -tags "$(BUILD_TAGS)" $(GO_TEST_FLAGS)
+	KUBEBUILDER_ASSETS="$(KUBEBUILDER_ASSETS)" $(GO_TEST) $(UNIT_TEST_PACKAGES) -cover -tags "$(BUILD_TAGS)" $(GO_TEST_FLAGS) -ldflags "-X github.com/aws/eks-anywhere/pkg/version.gitVersion=v0.0.0-dev"
```
Comment (cxbrowne1207) [inline Makefile:425]: What is the purpose of this change? Is this something we want to commit into the repo, or a change made for testing purposes?

File: pkg/api/v1alpha1/cluster_types.go [hunk 2]
```diff
@@ -63,9 +63,28 @@ type ClusterSpec struct {
...
+		return true
+	}
+	if n == nil || o == nil {
+		return false
+	}
+	return *n == *o
+}
+
+// DevEksaVersion can be used in tests.
+var DevEksaVersion EksaVersion = "v0.0.0-dev"
```
Comment (cxbrowne1207) [inline pkg/api/v1alpha1/cluster_types.go:86]: Can we move this somewhere in the `internal/test` folder instead then?

File: pkg/api/v1alpha1/cluster_types.go [hunk 3]
```diff
@@ -1227,14 +1249,23 @@ func (c *Cluster) ConvertConfigToConfigGenerateStruct() *ClusterGenerate {
...
 
+// ManagedBy returns the Cluster's management cluster's name.
 func (c *Cluster) ManagedBy() string {
 	return c.Spec.ManagementCluster.Name
 }
 
+// EKSAReleaseName maps EksaVersion to an EKSARelease.
+func (c *Cluster) EKSAReleaseName() string {
+	name := "eksa-"
+	version := string(*c.Spec.EksaVersion)
```
Comment (jiayiwang7) [inline pkg/api/v1alpha1/cluster_types.go:1265]: this throws error if EksaVersion is nil

File: pkg/api/v1alpha1/cluster_types.go [hunk 4]
```diff
@@ -1227,14 +1249,23 @@ func (c *Cluster) ConvertConfigToConfigGenerateStruct() *ClusterGenerate {
...
+	return name + strings.ReplaceAll(version, ".", "-")
```
Comment (jiayiwang7) [inline pkg/api/v1alpha1/cluster_types.go:1266]:
```suggestion
	return fmt.Sprintf("eksa-%s", strings.ReplaceAll(version, ".", "-")
```

Comment (jiayiwang7) [reply pkg/api/v1alpha1/cluster_types.go:1266]: also do we need to validate the format of version here if not taken care of in somewhere else

File: pkg/api/v1alpha1/eksarelease_types.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// EKSARelease is the mapping between CLI version and Bundle.
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// Cluster is the Schema for the clusters API.
```
Comment (jiayiwang7) [inline pkg/api/v1alpha1/eksarelease_types.go:8]: remove?

**Review by jiayiwang7 — COMMENTED** [review]
don't forget to run `make release-manifests` to re-generate the eksa-components.yaml and related crd changes.

File: Makefile (hunk 1 above)
Comment (tatlat) [reply Makefile]: This is needed for unit tests. Admittedly, the unit tests that use version.gitVersion are not in this PR but it will be needed in related PRs.

File: pkg/api/v1alpha1/cluster_types.go (hunk 2 above)
Comment (tatlat) [reply pkg/api/v1alpha1/cluster_types.go:86]: yes

File: pkg/api/v1alpha1/cluster_types.go (hunk 4 above)
Comment (tatlat) [reply pkg/api/v1alpha1/cluster_types.go]: version would be validated before this is called

File: Makefile (hunk 1 above)
Comment (tatlat) [reply Makefile]: We don't need it. I removed it.

File: release/api/v1alpha1/eksarelease_types.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// EKSARelease is the mapping between release semver of EKS-A and a Bundles resource on the cluster.
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+type EKSARelease struct {
```
Comment (cxbrowne1207) [inline release/api/v1alpha1/eksarelease_types.go:8]: Do we not also need an `EKSAReleaseList` type?

Comment (tatlat) [reply release/api/v1alpha1/eksarelease_types.go:8]: Yes, I added it.

File: release/api/v1alpha1/eksarelease_types.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   EKSAReleaseSpec   `json:"spec,omitempty"`
+	Status EKSAReleaseStatus `json:"status,omitempty"`
```
Comment (g-gaston) [inline release/api/v1alpha1/eksarelease_types.go:13]: do we need a status?

File: config/rbac/role.yaml [hunk 8]
```diff
@@ -49,6 +49,14 @@ rules:
   - patch
   - update
   - watch
+- apiGroups:
```
Comment (g-gaston) [inline config/rbac/role.yaml:52]: did you update this manually? we should do it through kubebuilder annotations
Also, could we scope this down to eksa-system ns?

File: release/api/v1alpha1/eksarelease_types.go (hunk 7 above)
Comment (tatlat) [reply release/api/v1alpha1/eksarelease_types.go]: no, i removed it

File: controllers/cluster_controller.go [hunk 9]
```diff
@@ -201,6 +201,7 @@ func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager, log logr.Logger)
 // +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=awssnowclusters;awssnowmachinetemplates;awssnowippools;vsphereclusters;vspheremachinetemplates;dockerclusters;dockermachinetemplates;tinkerbellclusters;tinkerbellmachinetemplates;cloudstackclusters;cloudstackmachinetemplates;nutanixclusters;nutanixmachinetemplates,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=packages.eks.amazonaws.com,resources=packages,verbs=create;delete;get;list;patch;update;watch
 // +kubebuilder:rbac:groups=packages.eks.amazonaws.com,namespace=eksa-packages,resources=packagebundlecontrollers,verbs=delete
+// +kubebuilder:rbac:groups=anywhere.eks.amazonaws.com,namespace=eksa-packages,resources=eksareleases,verbs=get;list;watch
```
Comment (g-gaston) [inline controllers/cluster_controller.go:204]: this won't work, we are not gonna install the eksa release object in the eksa-packages namespace

File: controllers/cluster_controller.go [hunk 10]
```diff
@@ -180,7 +180,7 @@ func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager, log logr.Logger)
 
 // +kubebuilder:rbac:groups="",resources=events,verbs=create;patch;update
 // +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;delete
-// +kubebuilder:rbac:groups="",namespace=eksa-packages,resources=secrets,verbs=patch;update
+// +kubebuilder:rbac:groups="",namespace=eksa-system,resources=secrets,verbs=patch;update
```
Comment (ddjjia) [inline controllers/cluster_controller.go:183]: thanks for correcting the namespace!

File: config/rbac/role.yaml [hunk 11]
```diff
@@ -260,7 +260,7 @@ kind: Role
 metadata:
   creationTimestamp: null
   name: manager-role
-  namespace: eksa-packages
```
[the remark this answers is not in the document]
Comment (ddjjia) [reply config/rbac/role.yaml:263]: Our `make generate-manifests` generates role under namespace as what specified in annotation. but release manifests always has role under eksa-system. Tanvir and I verified last Friday the actual role in the cluster is eksa-system namespaced and we ran create/scale/lable/taints/delete tests and make sure it worked

File: controllers/cluster_controller.go [hunk 12]
```diff
@@ -200,7 +200,8 @@ func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager, log logr.Logger)
 // +kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=machines,verbs=list;watch
 // +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=awssnowclusters;awssnowmachinetemplates;awssnowippools;vsphereclusters;vspheremachinetemplates;dockerclusters;dockermachinetemplates;tinkerbellclusters;tinkerbellmachinetemplates;cloudstackclusters;cloudstackmachinetemplates;nutanixclusters;nutanixmachinetemplates,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=packages.eks.amazonaws.com,resources=packages,verbs=create;delete;get;list;patch;update;watch
-// +kubebuilder:rbac:groups=packages.eks.amazonaws.com,namespace=eksa-packages,resources=packagebundlecontrollers,verbs=delete
+// +kubebuilder:rbac:groups=packages.eks.amazonaws.com,namespace=eksa-system,resources=packagebundlecontrollers,verbs=delete
```
[the remark this answers is not in the document]
Comment (cxbrowne1207) [reply controllers/cluster_controller.go:203]: Seems like this should be in the `eksa-packages `namespace

[the remark this answers is not in the document]
Comment (cxbrowne1207) [reply controllers/cluster_controller.go:203]: Seems like this should be in the `eksa-packages` namespace
